File: back/src/main/java/csh/back/domain/vote/vote/service/VoteService.java
package csh.back.domain.vote.vote.service;

import csh.back.domain.trip.group.entity.TripGroup;
import csh.back.domain.trip.group.repository.TripGroupRepository;
import csh.back.domain.trip.member.entity.TripMember;
import csh.back.domain.trip.member.repository.TripMemberRepository;
import csh.back.domain.trip.member.validator.TripMemberValidator;
import csh.back.domain.trip.place.dto.response.TripPlaceFindResponse;
import csh.back.domain.trip.place.service.TripPlaceService;
import csh.back.domain.trip.timeline.entity.Timeline;
import csh.back.domain.trip.timeline.repository.TimelineRepository;
import csh.back.domain.vote.item.entity.VoteItem;
import csh.back.domain.vote.item.repository.VoteItemRepository;
import csh.back.domain.vote.user.entity.VoteUser;
import csh.back.domain.vote.user.repository.VoteUserRepository;
import csh.back.domain.vote.vote.dto.response.*;
import csh.back.domain.vote.vote.dto.web.VoteTimelineResponse;
import csh.back.domain.vote.vote.entity.Vote;
import csh.back.domain.vote.vote.enums.VoteStatus;
import csh.back.domain.vote.vote.repository.VoteRepository;
import csh.back.domain.vote.vote.repository.VoteTimelineIdProjection;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class VoteService {
    private final VoteRepository voteRepository;
    private final VoteItemRepository voteItemRepository;
    private final VoteUserRepository voteUserRepository;
    private final TripGroupRepository tripGroupRepository;
    private final TimelineRepository timeLineRepository;
    private final TripMemberValidator tripMemberValidator;
    private final TripMemberRepository tripMemberRepository;
    private final TripPlaceService tripPlaceService;

    private final int DEFAULT_UPDATE_COUNT = 0;

    // 투표 목록 조회해오는거(투표탭에서 사용됨)
    public List<VoteFindListResponse> findVoteList(Long tripGroupId, Long memberId) {
        tripMemberValidator.validMember(tripGroupId, memberId);

        TripGroup tripGroup = tripGroupRepository.findById(tripGroupId).orElseThrow(RuntimeException::new);

        Integer totalDays = tripGroup.getNights() + 1;
        Map<Long, List<Timeline>> byDay = timeLineRepository.findAllByTripGroupId(tripGroupId).stream()
                .collect(Collectors.groupingBy(Timeline::getDayNumber));

        Map<Long, Vote> byTimeLineId = voteRepository.findVotesWithTimelineByTripGroupId(tripGroupId).stream()
                .collect(Collectors.toMap(
                        vote -> vote.getTimeline().getId(),
                        vote -> vote
                ));

        List<VoteFindListResponse> voteFindListResponses = new ArrayList<>();
        for (int day = 1; day <= totalDays; day++) {
            List<VoteWithTimelineResponse> timeLineResponses =
                    byDay.getOrDefault((long) day, List.of()).stream()
                            .map(timeLine -> createVoteAndTimeLineResponse(timeLine, byTimeLineId))
                            .toList();

            voteFindListResponses.add(
                    VoteFindListResponse.of(tripGroup.getStartDate().plusDays(day - 1), timeLineResponses)
            );
        }

        return voteFindListResponses;
    }

    // 장소 : 몇표를 표현하기 위한 메소드(투표하는 화면 진입 시 사용)
    public VoteFindWithUpdateCountResponse findVoteItemAndCount(Long tripGroupId, Long voteId, Long memberId) {
        tripMemberValidator.validMember(tripGroupId, memberId);

        TripMember tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId).orElseThrow(RuntimeException::new); // 이게 없으면 에러가 맞지
        VoteUser voteUser = voteUserRepository.findByVoteIdAndTripMemberId(voteId ,tripMember.getId()).orElse(null); // 이건 없을수있지

        //투표된 장소 목록을 조회해 옴
        List<VoteItem> voteItemList = voteItemRepository.findAllByVoteIdWithTripPlace(voteId);
        //각 장소에 몇포가 투표 되었는지 카운팅
        Map<Long, Long> countMap = voteCount(voteId);

        Vote vote = voteRepository.findById(voteId).orElseThrow(RuntimeException::new);

        VoteItem voteItem = voteUser != null ? voteUser.getVoteItem() : null;
        int updateCount = voteUser != null ? voteUser.getUpdateCount() : DEFAULT_UPDATE_COUNT;

        List<VoteFindResponse> voteFindResponses = voteItemList.stream()
                .map(vi -> VoteFindResponse.of(
                        vi.getTripPlace().getId(),
                        vi.getTripPlace().getName(),
                        countMap.getOrDefault(vi.getId(), 0L),
                        vi.equals(voteItem)
                ))
                .toList();
        List<TripPlaceFindResponse> wishPlaceFindResponses = tripPlaceService.findWishPlaces(tripGroupId, memberId);
        //장소의 아이디를 키로 하여 위의 맵에서 횟수를 매핑하여 반환
        return VoteFindWithUpdateCountResponse.of(voteFindResponses, wishPlaceFindResponses, updateCount, vote);
    }

    public Map<Long, Long> findAllVoteIds(List<Long> timeLineIds) {
        List<VoteTimelineIdProjection> voteTimeLineIdProjections = voteRepository.findVoteIdsByTimeLineIds(timeLineIds);
        return voteTimeLineIdProjections.stream()
                .collect(Collectors.toMap(
                    item -> item.getTimelineId(),
                    item -> item.getVoteId()
                ));
    }

    // 해당 장소에 투표한 유저가 누구인지 표현하기 위한 메소드(현재 사용안됨)
    public List<VoteFindUserResponse> findUserVoteThisPlace(Long tripGroupId, Long voteId, Long tripPlaceId, Long memberId) {
        tripMemberValidator.validMember(tripGroupId, memberId);

        VoteItem voteItem = voteItemRepository.findByVoteIdAndTripPlaceId(voteId, tripPlaceId).orElseThrow(RuntimeException::new);
        List<VoteUser> voteUsers = voteUserRepository.findByVoteItemId(voteItem.getId());
        List<VoteFindUserResponse> responses = voteUsers.stream().map(VoteFindUserResponse::from).toList();
        return responses;
    }

    @Transactional
    public VoteCreateResponse wrapperCreateVote(Long tripGroupId, Long memberId, Long timeLineId) {
        Timeline timeLine = timeLineRepository.findById(timeLineId).orElseThrow(RuntimeException::new);
        return createVote(tripGroupId, memberId, timeLine);
    }

    @Transactional
    public VoteCreateResponse createVote(Long tripGroupId, Long memberId, Timeline timeLine) {
        tripMemberValidator.validMember(tripGroupId, memberId);
        TripGroup tripGroup = tripGroupRepository.findById(tripGroupId).orElseThrow(RuntimeException::new);
        TripMember tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId).orElseThrow(RuntimeException::new);
        LocalDateTime expireTime = tripGroup.getStartDate().minusDays(1).atStartOfDay();
        Vote vote = Vote
                .builder()
                .tripGroup(tripGroup)
                .timeline(timeLine)
                .tripMember(tripMember)
                .expireTime(expireTime)
                .build();
        Vote saved = voteRepository.save(vote);
        return VoteCreateResponse.from(saved);
    }

    @Transactional
    public void createVoteBatch(Long tripGroupId, Long memberId, List<Timeline> timeLines) {
        tripMemberValidator.validMember(tripGroupId, memberId);
        TripGroup tripGroup = tripGroupRepository.findById(tripGroupId).orElseThrow(RuntimeException::new);
        LocalDateTime expireTime = tripGroup.getStartDate().minusDays(1).atStartOfDay();
        TripMember tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId).orElseThrow(RuntimeException::new);

        List<Vote> votes = timeLines.stream()
                .map(timeLine -> Vote
                        .builder()
                        .tripGroup(tripGroup)
                        .timeline(timeLine)
                        .tripMember(tripMember)
                        .expireTime(expireTime)
                        .build())
                .toList();
        voteRepository.saveAll(votes);
    }

    public VoteTimelineResponse voteConfirm(Long maxVoteItemId, Long voteId) {
        VoteItem voteItem = voteItemRepository.findById(maxVoteItemId).orElseThrow(RuntimeException::new);
        Long confirmPlaceId = voteItem.getTripPlace().getId();
        Timeline timeLine = voteRepository.findTimeLineByVoteId(voteId).orElseThrow(RuntimeException::new);
        voteItem.getVote().updateStatus(VoteStatus.CONFIRMED);
        return VoteTimelineResponse.of(confirmPlaceId, timeLine);
    }

    public Map<Long, Long> voteCount(Long voteId) {
        return voteUserRepository.countGroupByVoteId(voteId)
                .stream()
                .collect(Collectors.toMap(
                        row -> row.getVoteItemId(),
                        row -> row.getVoteCount()
                ));
    }

    public void expireVote(Long voteId) {
        Vote vote = voteRepository.findById(voteId).orElseThrow(RuntimeException::new);
        vote.updateStatus(VoteStatus.EXPIRED);
    }

    private VoteWithTimelineResponse createVoteAndTimeLineResponse(
            Timeline timeLine,
            Map<Long, Vote> byTimeLindId
    ) {
        Long timeLineId = timeLine.getId();
        Vote vote = byTimeLindId.get(timeLineId);
        if(vote == null) {
            log.error("Timeline {}과 연결된 Vote가 존재 하지 않습니다! 확인 해주세요!", timeLineId);
            return VoteWithTimelineResponse.of(timeLine, null);
        }
        return VoteWithTimelineResponse.of(timeLine, vote);
    }


}

File: back/src/main/kotlin/csh/back/domain/vote/vote/service/VoteService.kt
package csh.back.domain.vote.vote.service

import csh.back.domain.trip.chat.service.ChatService
import csh.back.domain.trip.event.dto.TripEvent
import csh.back.domain.trip.event.enums.TripEventType
import csh.back.domain.trip.event.service.TripEventService
import csh.back.domain.trip.group.entity.TripGroup
import csh.back.domain.trip.group.exception.NonMemberException
import csh.back.domain.trip.group.exception.NotFoundException
import csh.back.domain.trip.group.repository.TripGroupRepository
import csh.back.domain.trip.group.settings.repository.TripGroupSettingsRepository
import csh.back.domain.trip.member.repository.TripMemberRepository
import csh.back.domain.trip.place.service.TripPlaceService
import csh.back.domain.trip.timeline.entity.Timeline
import csh.back.domain.trip.timeline.repository.TimelineRepository
import csh.back.domain.vote.item.repository.VoteItemRepository
import csh.back.domain.vote.user.repository.VoteUserRepository
import csh.back.domain.vote.vote.dto.response.VoteCreateResponse
import csh.back.domain.vote.vote.dto.response.VoteFindListResponse
import csh.back.domain.vote.vote.dto.response.VoteFindResponse
import csh.back.domain.vote.vote.dto.response.VoteFindUserResponse
import csh.back.domain.vote.vote.dto.response.VoteFindWithUpdateCountResponse
import csh.back.domain.vote.vote.dto.response.VoteWithTimelineResponse
import csh.back.domain.vote.vote.dto.response.VoterResponse
import csh.back.domain.vote.vote.dto.web.VoteTimelineResponse
import csh.back.domain.vote.vote.entity.Vote
import csh.back.domain.vote.vote.enums.VoteStatus
import csh.back.domain.vote.vote.repository.VoteRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class VoteService(
    private val voteRepository: VoteRepository,
    private val voteItemRepository: VoteItemRepository,
    private val voteUserRepository: VoteUserRepository,
    private val tripGroupRepository: TripGroupRepository,
    private val tripGroupSettingsRepository: TripGroupSettingsRepository,
    private val timelineRepository: TimelineRepository,
    private val tripMemberRepository: TripMemberRepository,
    private val tripPlaceService: TripPlaceService,
    private val tripEventService: TripEventService,
    private val chatService: ChatService,
) {

    fun findVoteList(tripGroupId: Long, memberId: Long): List<VoteFindListResponse> {
        val tripGroup = validateTripMember(tripGroupId, memberId)
        val totalDays = tripGroup.nights + 1
        val byDay = timelineRepository.findAllByTripGroupId(tripGroupId)
            .groupBy { it.dayNumber }
        val byTimelineId = voteRepository.findVotesWithTimelineByTripGroupId(tripGroupId)
            .associateBy { it.timeline.id }

        val voteFindListResponses = mutableListOf<VoteFindListResponse>()
        for (day in 1..totalDays) {
            val timelineResponses = byDay.getOrDefault(day.toLong(), emptyList())
                .map { timeline -> createVoteAndTimelineResponse(timeline, byTimelineId) }
            voteFindListResponses.add(
                VoteFindListResponse.of(tripGroup.startDate.plusDays((day - 1).toLong()), timelineResponses),
            )
        }
        return voteFindListResponses
    }

    fun findVoteItemAndCount(tripGroupId: Long, voteId: Long, memberId: Long): VoteFindWithUpdateCountResponse {
        validateTripMember(tripGroupId, memberId)
        val tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId)
            .orElseThrow(::RuntimeException)
        val voteUser = voteUserRepository.findByVoteIdAndTripMemberId(voteId, tripMember.id!!).orElse(null)
        val voteItemList = voteItemRepository.findAllByVoteIdWithTripPlace(voteId)
        val countMap = voteCount(voteId)
        val vote = voteRepository.findById(voteId).orElseThrow(::RuntimeException)
        val votersByVoteItemId: Map<Long?, List<VoterResponse>> = if (vote.isAnonymous) {
            emptyMap()
        } else {
            voteUserRepository.findAllByVoteIdWithVoter(voteId)
                .groupBy({ it.voteItem.id }, { VoterResponse(it.tripMember.id!!, it.tripMember.member.name) })
        }
        val voteItem = voteUser?.voteItem
        val updateCount = voteUser?.updateCount ?: DEFAULT_UPDATE_COUNT
        val voteFindResponses = voteItemList.map { vi ->
            VoteFindResponse.of(
                vi.tripPlace.id,
                vi.tripPlace.name,
                countMap.getOrDefault(vi.id, 0L),
                vi == voteItem,
                votersByVoteItemId.getOrDefault(vi.id, emptyList()),
            )
        }
        val wishPlaceFindResponses = tripPlaceService.findWishPlaces(tripGroupId, memberId)
        return VoteFindWithUpdateCountResponse.of(voteFindResponses, wishPlaceFindResponses, updateCount, vote)
    }

    fun findAllVoteIds(timelineIds: List<Long>): Map<Long, Long> {
        val voteTimelineIdProjections = voteRepository.findVoteIdsByTimelineIds(timelineIds)
        return voteTimelineIdProjections.associate { it.getTimelineId() to it.getVoteId() }
    }

    fun findUserVoteThisPlace(tripGroupId: Long, voteId: Long, tripPlaceId: Long, memberId: Long): List<VoteFindUserResponse> {
        validateTripMember(tripGroupId, memberId)
        val voteItem = voteItemRepository.findByVoteIdAndTripPlaceId(voteId, tripPlaceId).orElseThrow(::RuntimeException)
        val voteUsers = voteUserRepository.findByVoteItemId(voteItem.id!!)
        return voteUsers.map(VoteFindUserResponse::from)
    }

    @Transactional
    fun wrapperCreateVote(tripGroupId: Long, memberId: Long, timelineId: Long): VoteCreateResponse {
        val timeline = timelineRepository.findById(timelineId).orElseThrow(::RuntimeException)
        val response = createVote(tripGroupId, memberId, timeline)
        tripEventService.publishAfterCommit(
            TripEvent(
                eventType = TripEventType.VOTE_CREATED,
                message = "${timeline.dayNumber}일차 시간 구간 투표 생성",
                tripGroupId = tripGroupId,
                actorMemberId = memberId,
                dayNumber = timeline.dayNumber,
                timelineId = timelineId,
                voteId = requireNotNull(response.voteId),
            ),
        )
        chatService.recordSystemMessage(tripGroupId, "${timeline.dayNumber}일차 시간 구간 투표 생성")
        return response
    }

    @Transactional
    fun createVote(tripGroupId: Long, memberId: Long, timeline: Timeline): VoteCreateResponse {
        val tripGroup = validateTripMember(tripGroupId, memberId)
        val tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId).orElseThrow(::RuntimeException)
        val expireTime = tripGroup.startDate.minusDays(1).atStartOfDay()
        val isAnonymous = resolveIsAnonymous(tripGroupId)
        val vote = Vote(tripGroup = tripGroup, timeline = timeline, tripMember = tripMember, expireTime = expireTime, isAnonymous = isAnonymous)
        val saved = voteRepository.save(vote)
        return VoteCreateResponse.from(saved)
    }

    @Transactional
    fun createVoteBatch(tripGroupId: Long, memberId: Long, timelines: List<Timeline>) {
        val tripGroup = validateTripMember(tripGroupId, memberId)
        val expireTime = tripGroup.startDate.minusDays(1).atStartOfDay()
        val tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId).orElseThrow(::RuntimeException)
        val isAnonymous = resolveIsAnonymous(tripGroupId)
        val votes = timelines.map { timeline ->
            Vote(tripGroup = tripGroup, timeline = timeline, tripMember = tripMember, expireTime = expireTime, isAnonymous = isAnonymous)
        }
        voteRepository.saveAll(votes)
    }

    private fun resolveIsAnonymous(tripGroupId: Long): Boolean =
        tripGroupSettingsRepository.findByTripGroupId(tripGroupId).map { it.isAnonymousVote }.orElse(true)

    @Transactional
    fun lockPendingVote(voteId: Long): Vote {
        val vote = voteRepository.findByIdWithLock(voteId).orElseThrow(::RuntimeException)
        check(vote.status == VoteStatus.PENDING) { "이미 확정되었거나 만료된 투표입니다." }
        return vote
    }

    @Transactional
    fun updateAnonymous(tripGroupId: Long, voteId: Long, memberId: Long, isAnonymous: Boolean): Boolean {
        validateTripAdmin(tripGroupId, memberId)
        val vote = voteRepository.findById(voteId).orElseThrow(::RuntimeException)
        check(vote.status == VoteStatus.PENDING) { "이미 확정되었거나 만료된 투표입니다." }
        vote.updateAnonymous(isAnonymous)
        return vote.isAnonymous
    }

    private fun validateTripAdmin(tripGroupId: Long, memberId: Long) {
        val isAdmin = tripMemberRepository.existsByTripGroupIdAndMemberIdAndIsAdminTrue(tripGroupId, memberId)
        require(isAdmin) { "여행 모임 방장만 접근할 수 있습니다." }
    }

    fun voteConfirm(maxVoteItemId: Long, voteId: Long): VoteTimelineResponse {
        val voteItem = voteItemRepository.findById(maxVoteItemId).orElseThrow(::RuntimeException)
        val confirmPlaceId = voteItem.tripPlace.id!!
        val timeline = voteRepository.findTimelineByVoteId(voteId).orElseThrow(::RuntimeException)
        voteItem.vote.updateStatus(VoteStatus.CONFIRMED)
        return VoteTimelineResponse.of(confirmPlaceId, timeline)
    }

    fun voteCount(voteId: Long): Map<Long, Long> =
        voteUserRepository.countGroupByVoteId(voteId).associate { row -> row.getVoteItemId() to row.getVoteCount() }

    fun expireVote(voteId: Long) {
        val vote = voteRepository.findById(voteId).orElseThrow(::RuntimeException)
        vote.updateStatus(VoteStatus.EXPIRED)
    }

    private fun createVoteAndTimelineResponse(timeline: Timeline, byTimelineId: Map<Long?, Vote>): VoteWithTimelineResponse {
        val timelineId = timeline.id
        val vote = byTimelineId[timelineId]
        if (vote == null) {
            log.error("Timeline {}과 연결된 Vote가 존재 하지 않습니다! 확인 해주세요!", timelineId)
            return VoteWithTimelineResponse.of(timeline, null)
        }
        return VoteWithTimelineResponse.of(timeline, vote)
    }

    private fun validateTripMember(tripGroupId: Long, memberId: Long): TripGroup {
        val tripGroup = tripGroupRepository.findById(tripGroupId)
            .orElseThrow {
                NotFoundException("존재하지 않는 모임입니다.")
            }

        if (!tripMemberRepository.existsByTripGroupIdAndMemberId(tripGroupId, memberId)) {
            throw NonMemberException("해당 모임의 멤버가 아닙니다.")
        }

        return tripGroup
    }

    companion object {
        private val log = LoggerFactory.getLogger(VoteService::class.java)
        private const val DEFAULT_UPDATE_COUNT = 0
    }
}
